runtime/lua/vim/lsp/util.lua

```diff
@@ -683,6 +684,52 @@ do
       api.nvim_buf_set_virtual_text(bufnr, diagnostic_ns, line, virt_texts, {})
     end
   end
+  function M.buf_diagnostics_statusline(bufnr, diagnostics)
+    all_buffer_diagnostics_counts[bufnr] = { errors=0, warnings=0, hints=0, info=0 }
+    for _, diagnostic in ipairs(diagnostics) do
+      if diagnostic.severity == 1 then
+        all_buffer_diagnostics_counts[bufnr]['errors'] = all_buffer_diagnostics_counts[bufnr]['errors'] + 1
+      elseif diagnostic.severity == 2 then
+        all_buffer_diagnostics_counts[bufnr]['warnings'] = all_buffer_diagnostics_counts[bufnr]['warnings'] + 1
+      elseif diagnostic.severity == 3 then
+        all_buffer_diagnostics_counts[bufnr]['info'] = all_buffer_diagnostics_counts[bufnr]['info'] + 1
+      elseif diagnostic.severity == 4 then
+        all_buffer_diagnostics_counts[bufnr]['hints'] = all_buffer_diagnostics_counts[bufnr]['hints'] + 1
+      end
+    end
+  end
+  function M.buf_diagnostics_error_count()
+      local bufnr = vim.api.nvim_get_current_buf()
+      if all_buffer_diagnostics_counts[bufnr] ~= nil then
+          return all_buffer_diagnostics_counts[bufnr]['errors']
+      else
+          return '-'
+      end
+  end
+  function M.buf_diagnostics_warning_count()
+      local bufnr = vim.api.nvim_get_current_buf()
+      if all_buffer_diagnostics_counts[bufnr] ~= nil then
+          return all_buffer_diagnostics_counts[bufnr]['warnings']
+      else
+          return '-'
```

Comment:
why return a different type (instead of number)? And why would it be hyphen instead of empty string?

Reply:
The reason was to render `E: - W: -` when there was no diagnostic information, rather than saying that there are 0 errors and 0 warnings which may not be true. Should I return 0 instead?